src/models/helpers/genericCollaborator.js

```diff
@@ -0,0 +1,378 @@
+
+  // Find all entity CollaboratorType records that meet the specified criteria
+  const currentCollaboratorsForType = await sequelize.models[
+    collaboratorDetails[genericCollaboratorType].collaborators
+  ].findAll({
+    where: {
+      [[collaboratorDetails[genericCollaboratorType].idName]]: entityId,
+      linkBack: { [Op.contains]: filteredLinkBack },
+    },
+    include: [
+      {
+        model: sequelize.models.CollaboratorType,
+        as: 'collaboratorType',
+        required: true,
+        attributes: [],
+        where: {
+          name: typeName,
+        },
+        include: [{
+          model: sequelize.models.ValidFor,
+          as: 'validFor',
+          required: true,
+          attributes: [],
+          where: { name: collaboratorDetails[genericCollaboratorType].validFor },
+        }],
+      },
+    ],
+    ...(transaction && { transaction }),
+  });
+
+  if (currentCollaboratorsForType) {
+    // Separate the updates and deletes based on the conditions
+    const {
+      updates,
+      deletes,
+    } = currentCollaboratorsForType.reduce((acc, current) => {
+      if (current.dataValues.linkBack[linkBackKey].length > 1) {
+        // Remove the specified linkBack values from the array
+        const newLinkBack = {
+          ...current.dataValues.linkBack,
+          [linkBackKey]: current.dataValues.linkBack[linkBackKey]
+            .filter((value) => !linkBackValues.includes(value)),
+        };
+        acc.updates.push({
+          id: current.dataValues.id,
+          linkBack: newLinkBack,
+        });
+      } else if (Object.keys(current.dataValues.linkBack).length > 1) {
+        // Remove the specified linkBack key from the object
+        const newLinkBack = current.dataValues.linkBack;
+        delete newLinkBack[linkBackKey];
+        acc.updates.push({
+          id: current.dataValues.id,
+          linkBack: newLinkBack,
+        });
+      } else {
+        // Add the ID to the deletes array
+        acc.deletes.push(current.dataValues.id);
+      }
+      return acc;
+    }, { updates: [], deletes: [] });
+
+    // Update the entity CollaboratorType records and delete the specified records
+    await Promise.all([
+      ...(updates.length > 0
+        ? updates.map(async (update) => sequelize.models[
+          collaboratorDetails[genericCollaboratorType].collaborators
+        ].update(
+          { linkBack: update.linkBack },
+          {
+            where: { id: update.id },
+            independentHooks: true,
+            ...(transaction && { transaction }),
+          },
+        ))
+        : [Promise.resolve()]),
+      (deletes.length > 0
+        ? sequelize.models[
+          collaboratorDetails[genericCollaboratorType].collaborators
+        ].destroy({
+          where: { id: deletes },
+          independentHooks: true,
+          ...(transaction && { transaction }),
+        })
+        : Promise.resolve()),
+    ]);
```

Comment:
For simplicity sake, can this be refactored? It's pretty tough to follow as-is with the ternaries wrapped in a spread.

Untested simplification:

```javascript
const updatePromises = updates.map(async (update) => {
  await sequelize.models[collaboratorDetails[genericCollaboratorType].collaborators].update(
    { linkBack: update.linkBack },
    {
      where: { id: update.id },
      independentHooks: true,
      ...(transaction && { transaction }),
    },
  );
});

const deletePromise = sequelize.models[collaboratorDetails[genericCollaboratorType].collaborators].destroy({
  where: { id: deletes },
  independentHooks: true,
  ...(transaction && { transaction }),
});

await Promise.all([...updatePromises, deletePromise]);
```